src/com/github/Kraken3/AFKPGC/BotDetector.java

```diff
@@ -0,0 +1,306 @@
+
+	public void run() {
+		if (topSuspects == null) {
+			topSuspects = new TreeMap<Integer, Suspect>();
+		}
+		if (reprieve == null) {
+			reprieve = new HashMap<UUID, Integer>();
+		} else {
+			// decrement reprieve, regardless of plugin's active status.
+			for (Iterator<Map.Entry<UUID, Integer>> i = reprieve.entrySet().iterator(); i.hasNext(); ) {
+				Map.Entry<UUID, Integer> entry = i.next();
+				Integer roundsLeft = entry.getValue() - 1;
+				if (roundsLeft <= 0) { // no more reprieve
+					AFKPGC.debug("Reprieve is up for ", entry.getKey());
+					i.remove();
+				} else {
+					AFKPGC.debug("Reprieve decremented for ", entry.getKey());
+					entry.setValue(roundsLeft);
+				}
+			}
+		}
+
+		if (!AFKPGC.enabled) {
+			return;
+		}
+		
+		currentTPS = TpsReader.getTPS();
+		AFKPGC.debug("Bot Detector Running, TPS is: ", currentTPS);
+		Map<UUID, LastActivity> lastActivities = LastActivity.lastActivities;
+		if (currentTPS < acceptableTPS) {
+			topSuspects.clear();
+			int smallestMovedDistance = minBaselineMovement;
+			// find new top suspects
+			for (Map.Entry<UUID, LastActivity> entry : lastActivities.entrySet()) {
+				UUID playerUUID = entry.getKey();
+				Player p = Bukkit.getPlayer(playerUUID);
+				/* according to the author of AFKGPC, there might be
+				 * inconsistencies in this list, so this additional
+				 * check is needed */
+				// TODO: See if inconsistencies might be thread-safeness related.
+				if (lastActivities.containsKey(playerUUID) && p != null) {
+					LastActivity la = entry.getValue();
+					la.loggedLocations.add(p.getLocation());
+					if (la.loggedLocations.size() >= maxLocations) {
+						if (la.loggedLocations.size() > maxLocations) {
+							la.loggedLocations.removeFirst();
+						}
+						// we tracking location even if on reprieve or immune, but that's it
+						if (!reprieve.containsKey(playerUUID) && !AFKPGC.immuneAccounts.contains(playerUUID)) {
+							int itWasntMeISwear = la.calculateMovementRadius();
+							if (itWasntMeISwear < smallestMovedDistance) {
+								smallestMovedDistance = itWasntMeISwear;
+								Player dirtyLiar = Bukkit.getPlayer(playerUUID);
+
+								topSuspects.put(itWasntMeISwear, new Suspect(
+										playerUUID, dirtyLiar.getName(), dirtyLiar.getLocation(),
+										la.evaluateBounds(relaxationFactor) ) );
+								AFKPGC.debug("Player ", playerUUID, " added as suspect (movement was ",
+										itWasntMeISwear, ")");
+
+								if (topSuspects.size() > maxSuspects) {
+									Suspect cleared = topSuspects.pollLastEntry().getValue(); // gets rid of largest distance
+									AFKPGC.debug("Player ", cleared.getUUID(), " released as suspect, better suspects found");
+								}
+							}
+						} else {
+							AFKPGC.debug("Skipping ", playerUUID, " due to reprieve or immunity");
+						}
+					} else {
+						AFKPGC.debug("Skipping ", playerUUID, " due to insufficient location data");

```

Comment:
Doesn't this just cause unnecessary spam? With 2 players on the test server it's probably okay, but if we test this on the main server with ~100 players, it would send 1000 messages before even starting to do something.